torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,268 @@
+      [:logical_and, :logical_or, :logical_xor] ++
+      [:outer]
+
+
+  for op <- binary_ops do
+    @impl true
+    def unquote(op)(out, left, right) when is_number(right) do
+      NIF.unquote(op)(left.data.ref, right) |> from_ref(out)
+    end
+
+    def unquote(op)(out, left, right) do
+      NIF.unquote(op)(left.data.ref, right.data.ref) |> from_ref(out)
+    end
+  end
+
+  unary_ops =
+    Enum.map(Nx.Shared.unary_math_funs(), &elem(&1, 0)) ++
+      [:abs, :bitwise_not, :ceil, :floor, :negate, :round, :sign]
+      # [:count_leading_zeros, :population_count]
+
+  for op <- unary_ops do
+    @impl true
+    def unquote(op)(out, tensor) do
+      NIF.unquote(op)(tensor.data.ref) |> from_ref(out)
+    end
+  end
+
+
+  @impl true
+  def dot(
+        out,
+        %T{data: %TB{ref: left_ref}},
+        _axes1,
+        %T{data: %TB{ref: right_ref}},
+        _axes2
+      ) do
+    NIF.dot(left_ref, right_ref) |> from_ref(out)
+  end
+
+  @impl true
+  def cholesky(%T{} = out, %T{data: %TB{ref: ref}}) do
+    NIF.cholesky(ref) |> from_ref(out)
+  end
+
+  @impl true
+  def qr(
+        {q_holder, r_holder},
+        tensor,
+        opts
+      ),
+      do: NIF.qr(tensor.data.ref, opts[:mode] == :reduced) |> from_pair_ref({q_holder, r_holder})
+
+  @big_tensor_threshold_bytes 10_000_000
+
+  @impl true
+  def inspect(%T{type: {_, elem_size}} = tensor, inspect_opts) do
+    alias Inspect.Algebra, as: IA
+
+    limit = if(inspect_opts.limit == :infinity, do: nil, else: inspect_opts.limit + 1)
+
+    result =
+      if on_cpu?(tensor) do
+        byte_size = nbytes(tensor)
+        byte_limit = limit && limit * div(elem_size, 8)
+
+        if min(byte_limit, byte_size) > @big_tensor_threshold_bytes do
+          "Torchx tensor is too large to inspect. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
+        else
+          binary = to_blob(tensor, limit)
+          Nx.Backend.inspect(tensor, binary, inspect_opts)
+        end
+      else
+        "Tensors on the GPU cannot be inspected. Explicitly transfer the tensor by calling Nx.backend_transfer/1"
+      end
+
+    IA.concat(["Torchx.Backend(#{device(tensor)})", IA.line(), result])
+  end
+
+  ## Helpers
+
+  defp unwrap!({:ok, result}), do: result
+
+  defp unwrap!({:error, error}),
+    do: raise(RuntimeError, "Torchx: " <> List.to_string(error))
+
+  defp from_ref(maybe_ref, t), do: maybe_ref |> unwrap!() |> to_tensor(t)
+
+  defp from_pair_ref(maybe_ref, {t1, t2}) do
+    {left, right} = unwrap!(maybe_ref)
+    {to_tensor(left, t1), to_tensor(right, t2)}
+  end
+
+  defp from_list_ref(maybe_ref, t),
+    do:
+      maybe_ref
+      |> unwrap!()
+      |> Enum.map(&to_tensor(&1, t))
+
+  defp to_tensor(ref, %{type: out_type} = t),
+    do: %T{t | data: %__MODULE__{ref: maybe_cast_type(ref, out_type)}}
+
+  defp maybe_cast_type(ref, type) do
+    if from_torch_type(unwrap!(NIF.type(ref))) != type do
+      NIF.to_type(ref, torch_type(type)) |> unwrap!()
```

Comment:
When is this type casting supposed to happen? I think it would be better to not do this automatically... and only call maybe_cast_type on the few cases (for example, directly inside `def add` or wherever else expect them  we identify them to be different). Thoughts?

Reply:
This was added to always match with the `out` tensor expected type.
After changing type merging policies we still have some mismatches with PyTorch type merging. There are two cases for now:
1. Bitwise ops. PyTorch outputs the smallest type (e.g. `xor(s16, s64) == s16)`, and we promote to `s64`.
2. Power. PyTorch always promotes to `f64`, while we stay within general rule (e.g. `power(s16, s32) == s32`).
In both cases PyTorch policy makes sense, because we should throw away excess bits, if operands sizes do not match in bitwise. And in power we can have neg values in right operand, thus getting fractions in the result.
I agree with you, that we should add explicit type cast in these two cases.

Reply:
But wait, we should not do type casts in these cases, but change the `out` tensor expected type instead!:)